CDC logical replication overview (#3)
vaibhav-yb/yugabyte-db: docs/content/preview/explore/change-data-capture/using-logical-replication/overview.md

- LSN

  - An unsigned 64bit integer that uniquely identifies a change record or a transaction boundary record that is consumed from a given replication slot. In YugabyteDB, LSN values from different slots are considered unrelated and should not be compared. 

- Replica Identity

  - Replica identity is a table level parameter that can be used to control the amount of information being written to the change records. ([PG documentation](https://www.postgresql.org/docs/current/sql-altertable.html#SQL-ALTERTABLE-REPLICA-IDENTITY))

- Replication Protocols

  - Postgres has defined protocols for replication that need to be followed by clients to establish replication connection as well as message structures for streaming data. This includes two protocols - Streaming Replication protocol ([PG documentation](https://www.postgresql.org/docs/11/protocol-replication.html)), Logical Streaming Replication protocol ([PG documentation](https://www.postgresql.org/docs/11/protocol-logical-replication.html))

> The following sections explain each of the above concept in detail:

Author: The idea was to give a short definition for each concept and later expand on each of them. This way if a user just wants to get a rough idea, they can simply refer the definition. Incase they are interested to read more, they can refer the dedicated sections.

{{< note title="Note" >}}

Note: YugabyteDB does not support Physical Replication.

{{< /note >}}

{{< tip title="Explore" >}}
<!--TODO (Sumukh): Fix the Link to the getting started section. -->
See [Getting Started with Logical Replication](../../../explore/logical-replication/getting-started) in Explore to setup Logical Replication in YugabyteDB.

{{< /tip >}}

## **Guarantees**

|                                     |                                                                                                                                                                                                                                                                                                                                                                          |
| ----------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |

> | GUARANTEE                           | DESCRIPTION                                                                                                                                                                                                                                                                                                                                                              |

Reviewer: ```suggestion
| GUARANTEE | DESCRIPTION |
| :----- | :----- |
```

> | Per-slot ordered delivery guarantee | Changes from transactions from all the tables that are part of the replication slot’s publication are received in the order in which they were committed. This also implies ordered delivery across all the tablets that are part of the publication’s table list.                                                                                                       |

Reviewer: ```suggestion
| Per-slot ordered delivery guarantee | Changes from transactions from all the tables that are part of the replication slot’s publication are received in the order they were committed. This also implies ordered delivery across all the tablets that are part of the publication’s table list. |
```

> | At least once delivery              | Changes from transactions are streamed at least once. Changes from transactions may be streamed again in case of restart after failure. For example, this can happen in the case of a Kafka Connect Node failure. If the Kafka Connect Node pushes the records to Kafka and crashes before committing the offset, on restart, it will again get the same set of records. |

Reviewer: ```suggestion
| At least once delivery | Changes from transactions are streamed at least once. Changes from transactions may be streamed again in case of restart after failure. For example, this can happen in the case of a Kafka Connect node failure. If the Kafka Connect node pushes the records to Kafka and crashes before committing the offset, it will again get the same set of records upon restart. |
```

> | No gaps in change stream            | Receiving changes that are part of a transaction with commit time _t_ implies that you have already received changes from all transactions with commit time lower than _t_. Thus, receiving any change for a row with commit timestamp _t_,  implies that you have received all older changes for that row.                                                                     |

Reviewer: ```suggestion
| No gaps in change stream | Receiving changes that are part of a transaction with commit time *t* implies that you have already received changes from all transactions with commit time lower than *t*. Thus, receiving any change for a row with commit timestamp *t*,  implies that you have received all older changes for that row. |
```

## Highlights

### Resilience

> 1. Following a failure of the application or server or n/w, the replication can continue from any of the available server nodes 

Reviewer: ```suggestion
1. Following a failure of the application or server or network, the replication can continue from any of the available server nodes.
```

> 2. Replication continues from the transaction immediately after the transaction that was last acknowledged by the application. There will be no transactions that will be missed by the application. 

Reviewer: ```suggestion
2. Replication continues from the transaction immediately after the transaction that was last acknowledged by the application. There will be no transaction that will be missed by the application. 
```

### Security

>    CDC in YugabyteDB being based on the Postgres Logical Replication model means 

Reviewer: ```suggestion
CDC in YugabyteDB being based on the Postgres Logical Replication model means 
```

1. CDC user persona will be a 'PG Replication client'

2. A standard replication connection is used for consumption and all the server side configurations of authentication, authorizations, SSL modes and connection load balancing automatically can be leveraged

## Limitations

1. LSN Comparisons Across Slots

>     In the case of YugabyteDB, the LSN  does not represent the byte offset of a WAL record. Hence, arithmetic on LSN and any other usages of the LSN making this assumption will not work.Also, currently, comparison of LSN values from messages coming from different replication slots is not supported.

Reviewer: ```suggestion
    In the case of YugabyteDB, the LSN  does not represent the byte offset of a WAL record. Hence, arithmetic on LSN and any other usages of the LSN making this assumption will not work. Also, currently, comparison of LSN values from messages coming from different replication slots is not supported.
```

    
2. The functions listed below are currently unsupported:

>    - pg_current_wal_lsn
> 
>    - pg_wal_lsn_diff
> 
>    - IDENTIFY SYSTEM
> 
>    - txid_current
> 
>    - pg_stat_replication

Reviewer: ```suggestion
   - `pg_current_wal_lsn`
   - `pg_wal_lsn_diff`
   - `IDENTIFY SYSTEM`
   - `txid_current`
   - `pg_stat_replication`
```

>    Additionally, the functions responsible for pulling changes instead of the server streaming it are unsupported as well. They are described in the [Replication Functions section](https://www.postgresql.org/docs/11/functions-admin.html#FUNCTIONS-REPLICATION) in the PG documentation.

Reviewer: ```suggestion
   Additionally, the functions responsible for pulling changes instead of the server streaming it are unsupported as well. They are described in the [Replication Functions](https://www.postgresql.org/docs/11/functions-admin.html#FUNCTIONS-REPLICATION) section in the Postgres documentation.
```

   
3. Restriction on DDLs
    

>     DDL operations should not be performed from the time of replication slot creation till the start of snapshot consumption of the last table

Reviewer: ```suggestion
    DDL operations should not be performed from the time of replication slot creation till the start of snapshot consumption of the last table.
```

> 4. There should be a primary key on the table you want to stream the changes from

Reviewer: ```suggestion
4. There should be a primary key on the table you want to stream the changes from.
```

5. CDC is not supported on a target table for xCluster replication [11829](https://github.com/yugabyte/yugabyte-db/issues/11829).

6. Currently we don't support schema evolution for changes that require table rewrites (ex: ALTER TYPE).

7. YCQL tables aren't currently supported. Issue [11320](https://github.com/yugabyte/yugabyte-db/issues/11320).

> 

Reviewer: ```suggestion
```
